lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -1219,14 +1219,121 @@ struct CSEVisitor : NodeWalker {
+    auto hash = llvm::hash_value(V->getType());
+    auto &T = V->getPayload();
+    // Only use the first 8 elements in the hash. It's likely that if two
+    // tensors have different content they will diverge quickly. Fall back to
+    // full equality check in VarsEqDedup.
+    constexpr size_t maxNumEls = 8;
+    size_t numEls = std::min(T.getType().size(), maxNumEls);
+    size_t bufSize = T.getType().getElementSize() * numEls;
+    auto *data = T.getUnsafePtr();
+    for (size_t i = 0; i < bufSize; i++) {
+      hash = llvm::hash_combine(hash, data[i]);
+    }
+    return hash;
+  }
+};
+
+/// A helper type implementing the Variable equality predicate that can be used
+/// when Variable pointers are used as keys in hash maps for deduplication. It
+/// is assumed the Visibility and TrainKinds are the same, as deduplication
+/// only inserts if Private and None, respectively.
+struct VarsEqDedup {
+  bool operator()(const Variable *lhs, const Variable *rhs) const {
+    // Only consider Vars for deduplication if they have the same type. The
+    // train kind and visibility must already be the same.
+    if (lhs->getType() != rhs->getType()) {
+      return false;
+    }
+    assert(lhs->getVisibilityKind() == rhs->getVisibilityKind() &&
+           "Should only be comparing Variables with same VisibilityKind.");
+    assert(lhs->getTrainKind() == rhs->getTrainKind() &&
+           "Should only be comparing Variables with same TrainKind.");
+    // Only combine Vars if their data matches exactly, so allowed error is 0.0.
+    return lhs->getPayload().isEqual(rhs->getPayload(), /* allowedError */ 0.0);
+  }
+};
+
 } // namespace
 
+/// \returns true if Variable \p V is written into, either as a result, or as an
+/// overwritten input.
+static bool hasWriters(Variable *V) {
+  for (auto &U : V->getUsers()) {
+    auto *N = U.getUser();
+
+    // See if V is used as a result anywhere.
+    for (unsigned i = 0; i < N->getNumResults(); i++)
+      if (V == N->getNthResult(i))
+        return true;
+
+    // See if V is used as an overwritten input anywhere.
+    for (unsigned i = 0; i < N->getNumInputs(); i++)
+      if (N->isOverwrittenNthInput(i))
+        if (V == N->getNthInput(i))
+          return true;
+  }
+  return false;
+}
+
+/// Deduplicates constant variables in the Module \p M. Applicable constant
+/// variables for deduplication must have the same data, have
+/// VisibilityKind::Private, have TrainKind::None, and have no writers.
+static void deduplicateConstants(Module *M) {
+  // Map from Variables to other Variables that are equivalent for purposes of
+  // deduplication.
+  std::unordered_map<Variable *, Variable *, VarsHasherDedup, VarsEqDedup>
+      duplicateVars;
+
+  for (auto &V : M->getVars()) {
+    // Only perform deduplication on vars of small enough size. Otherwise just
+    // skip them. 256 is used as a heuristic, to keep compile time reasonable.
+    constexpr size_t maxNumEls = 256;
```

Comment:
Could we make this a cl::opt with the default being 256?
That way we can easily play with different sizes for tuning or if we have a bug we can easily rule that optimization out by putting maxNumEls to 0.

Reply:
Done.